### PR TITLE
Add `codspeed` benchmarks

### PR DESCRIPTION
**Pull request by gabrielmbmb**
## Description

This PR setups [CodSpeed](https://codspeed.io/dashboard) to run and profile benchmarks. Kudos to @plaguss for suggesting having benchmarks.

### CODE REVIEW

**Review by plaguss — APPROVED** [review]

**Comment by gabrielmbmb** [comment]
We can merge this PR after merging #615. `codspeed` requires `python>=3.12` for improved tracing (because in `3.12` the compatibility with the Linux perf profiler was added, and I think that's what `codspeed` uses behind the scenes)

**Pull request merged**
